c++/error-handling.md

```diff
@@ -0,0 +1,337 @@
+
+Contracts play an essential role in good API design.
+Most C++ APIs center around functions:
+expectations around their invocation,
+details of the types required by those invocations, and
+their supporting definitions and declarations.
+
+It is helpful to subcategorise API contracts in descending order of desirability:
+
+1. statically-enforceable;
+1. dynamically-enforceable; and
+1. manually-enforceable.
+
+Violation of statically-enforceable API contracts, e.g.
+
+```c++
+double sqrt(double a);
+
+void f()
+{
+  sqrt("4");
+}
+```
+
+and violation of manually-enforceable API contracts, e.g.
+
+```c++
+double sqrt(double a);
+
+// postconditions: returns x*x
+auto square(double x)
+{
+  return sqrt(x);
+}
+```
+
+Are outside the scope of this document.
+
+#### Dynamically-Enforceable API Contracts
+
+The provider should specify correct API usage.
+
+Example violation:
+
+```c++
+// precondition: a is non-negative
+double sqrt(double a);
+
+void f()
+{
+  sqrt(-1);
+}
+```
+
+* agreement: documentation, C++ Contracts (TBD)
+* provider: program author(s)
+* user: other program author(s)
+* _enforcement_: assertions, e.g. [`assert`](https://en.cppreference.com/w/cpp/error/assert)
+* violation: undefined behaviour (see discussion)
+
+## Dynamically-Enforceable Contracts
+
+Dynamically-Enforceable API Contracts and the ISO C++ Standard
+form a group of contracts whose violation represents run-time bugs.
+Of these, a subset can be expressed formally within the program
+using assertions (including API preconditions and postconditions).
+
+These expressions can be evaluated at run-time but typically incur a cost
+which violates the zero-overhead principle in the case of a bug-free program.
+
+Broadly, there are three possible strategies that can be adopted
+at the point where an assertion would evaluate to `false`:
+
+### Log-And-Continue Enforcement Strategy
+
+The assert statement could emit a run-time diagnostic regarding the contract violation
+and then continue past the assertion.
+
+This is unwise: the program is known to exhibit undefined behaviour at this
+point. If the provider were able to identify a reason why a violation might have
+occurred they would have identified a bug. If the provider had identified a bug,
+they should have fixed the bug. If the provider had fixed the bug,
+they would no longer be able to identify a reason why a violation might have occurred.
+Thus any violation is beyond the comprehension of the provider and "all bets are
+off" regarding the behaviour of the program. This is essentially what UB means.
+
+Nevertheless, under certain constraints
+
+### Trap Enforcement Strategy
+
+The assert statement can halt the program.
+
+This has the disadvantage that the program completely fails to perform its task.
+But the advantage is that no further incorrect behaviour
+will be exhibited by the program.
+
+### Prevention Enforcement Strategy
+
+Assuming bugs are eliminated, no assert expression will ever be false.
```

Comment:
> Assuming bugs are eliminated ...

I think, the core of the of the disagreement (whether to allow or disallow contract-based optimizations) lies in whether one can make this assumption. No amount of runtime testing (as opposed to a static correctness proof) can give a certainty that the program is bug-free. Also, even if we were able to run our program with all possible combinations of input and obtain the expected results, it is not a guarantee that there was no Abstract Machine Violation. The same binary that is later run on a minimally different architecture may expose different behavior.

I agree with your statement that only "bugs cause undefined behaviour", but there might be a HyrumBug situation at the abstract machine level: we have UB and the program works fine. But if we add contract-based optimizations to the mix, we break it.

Of course, an individual programmer (like a video game developer) may say, "I am willing to take the risk", and I am not arguing with this choice. But I am far from recommending it to the wider audience.

Reply:
I disagree with none of this and it's a central concern to all involved. However, one already cannot make this assumption in C++20. We already enable our optimisers, our code already has UB, and we survive.

Software is complex enough that it's no longer a matter of "100% bug free". It's a question of "how many 9's"?

Contracts +  UB + sanitizers + TDD gets you enough 9's. Log+continue loses 9's. Fuzz testing gets you a bunch more 9's.

> I agree with your statement that only "bugs cause undefined behaviour", but there might be a HyrumBug situation at the abstract machine level: we have UB and the program works fine. But if we add contract-based optimizations to the mix, we break it.

Yup. You don't do this unless you're foolhardy or you're serious about the 9's. I don't ever want to force anyone to do this. 

The main message I'd like to get across in this document is that while

1. `UB = faster + risk` when you enable the optimiser,
2. `UB = slower + certainty` when you enable the sanitizer.

If you only do 1) and not 2), yes the situation is worse with contract optimisations (which are optional). But if you do 2) and then 1), you should get safer, faster code.